README
Python-xrandr provides so called bindings for the xrandr extenstion of X.org,
the graphical base system found in most Linux and Unix systems. It is written
as a subproject of displayconfig-gtk and guidance, both graphical tools to
configure your screens and graphics.

Xrandr allows you to set the resolution and refresh rate of your monitor on the
fly. In the latest version 1.2 you can even configure multiple monitors 
instantly by merging or cloning monitors.

Currently only a subset of these capabilites are supported by python-xrandr.

The included command line tool pyxrandr shows you the detected devices and 
allows you to change the resolution, refresh rate and orientation or to 
disable an output.

Help on this project is always welcome. Feel free to contact us for questions
or comments.

The project is registered at Launchpad.net, a free (not as in beer) service 
infrastructure for Open Source software development hosted by Canonical, Ltd.

Errors (so called bugs) and problems can be reported here:

https://www.launchpad.net/python-xrandr/

Cheers,

Sebastian Heinlein <glatzor@ubuntu.com>, from the displayconfig-gtk team


debian/changelog
python-xrandr (0.1.1) gutsy; urgency=low

  * Fix a crash if the -dev packages are not installed. Thanks Alberto for
    reporting
  * Fix a too large screen size after disabling an output.

 -- Sebastian Heinlein <glatzor@ubuntu.com>  Tue, 27 Nov 2007 16:45:55 +0100

python-xrandr (0.1) gutsy; urgency=low

  * Initial Release.

 -- Sebastian Heinlein <glatzor@ubuntu.com>  Tue, 27 Nov 2007 09:39:08 +0100


tests/test_outputs.py
#!/usr/bin/python

import sys
sys.path.insert(0,"../")
from xrandr import xrandr

def check_crtc_for_output(screen, crtc):
    # FIXME: port the rest of check_crtc_for_output from xrandr.c
    # check if this is in use by some other output and ignore it then
    for other in screen.outputs.values():
        if other.id == output.id:
            continue
        if other.get_crtc() == crtc.xid:
            return False
    return True

if __name__ == "__main__":

    screen = xrandr.get_current_screen()

    output = screen.outputs["VGA"]
    print "output: ", output.id
    print "active: ", output.is_active()
    print "ctrc: ", output.get_crtc()
    print "ctrcs: ", output.get_crtcs()

    # pick a random mode
    modes = screen._resources.contents.modes
    mode = modes[0]
    print "Setting: ", mode.width, mode.height

    # find output that is connected to crtc
    # 
    # (why can't I just get from xrandr what output is connected to 
    # what crtc?!?)
    for crtc in output.get_crtcs():
        if check_crtc_for_output(screen, crtc):
            print "activating output %s" % output.id
            crtc.set_config(0,0, mode, [output])
    sys.exit(0)

    # diabling works too
    #screen.get_crtc_by_xid(output.get_crtc()).disable()
    

    # enlarge screen (VGA-0 starts where DVI-0 ends (right-of))
    # so that 2x1280 resolution is used
    # (this needs the "virtual" keyword in xorg.conf)
    new_res_x = 2*1280
    new_res_y = 1024
    fb_width_mm = screen.outputs["VGA-0"].get_physical_width() + screen.outputs["DVI-0"].get_physical_width()
    fb_height_mm = min(screen.outputs["VGA-0"].get_physical_height(),screen.outputs["DVI-0"].get_physical_height())
    screen.set_screen_size(new_res_x, new_res_y, fb_width_mm, fb_height_mm)
    new_output_x = new_res_x/2
    new_output_y = 0
    # normalize so that top-left is (0,0) (see xrandr.c:set_position)
    # ...
    for crtc in output.get_crtcs():
        if check_crtc_for_output(screen, crtc):
            print "enlarging output %s" % output.id
            crtc.set_config(new_output_x, new_output_y, mode, [output])
    


tests/test_outputs_renate.py
#!/usr/bin/python

import sys
sys.path.insert(0,"../")
from xrandr import xrandr

def check_crtc_for_output(screen, crtc):
    # FIXME: port the rest of check_crtc_for_output from xrandr.c
    # check if this is in use by some other output and ignore it then
    for other in screen.outputs.values():
        if other.id == output.id:
            continue
        if other.get_crtc() == crtc.xid:
            return False
    return True

if __name__ == "__main__":

    screen = xrandr.get_current_screen()

    output = screen.outputs["VGA"]
    print "output: ", output.id
    print "active: ", output.is_active()
    print "ctrc: ", output.get_crtc()
    print "ctrcs: ", output.get_crtcs()

    # pick a random mode
    modes = screen._resources.contents.modes
    mode = modes[0]
    print "Setting: ", mode.width, mode.height

    # find output that is connected to crtc
    # 
    # (why can't I just get from xrandr what output is connected to 
    # what crtc?!?)
    #for crtc in output.get_crtcs():
    #    if check_crtc_for_output(screen, crtc):
    #        print "activating output %s" % output.id
    #        crtc.set_config(0,0, mode, [output])
    #sys.exit(0)

    # diabling works too
    #screen.get_crtc_by_xid(output.get_crtc()).disable()
    

    # enlarge screen (VGA-0 starts where DVI-0 ends (right-of))
    # so that 2x1280 resolution is used
    # (this needs the "virtual" keyword in xorg.conf)
    new_res_x = 2*1024
    new_res_y = 768
    fb_width_mm = screen.get_output_by_name("VGA").get_physical_width() + screen.get_output_by_name("LVDS").get_physical_width()
    fb_height_mm = min(screen.get_output_by_name("VGA").get_physical_height(),screen.get_output_by_name("LVDS").get_physical_height())
    screen.set_size(new_res_x, new_res_y, fb_width_mm, fb_height_mm)
    new_output_x = new_res_x/2
    new_output_y = 0
    # normalize so that top-left is (0,0) (see xrandr.c:set_position)
    # ...
    for crtc in output.get_crtcs():
        if check_crtc_for_output(screen, crtc):
            print "enlarging output %s" % output.id
            crtc.set_config(new_output_x, new_output_y, mode, [output])
    


xrandr/gdk.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# Python-XRandR provides a high level API for the XRandR extension of the
# X.org server. XRandR allows to configure resolution, refresh rate, rotation 
# of the screen and multiple outputs of graphics cards.
#
# This module allows to get information for gtk.gdk.Screen objects
#
# In many aspects it follows the design of the xrand tool written by
# Keith Packard.
#
# Copyright 2007 © Sebastian Heinlein <sebastian.heinlein@web.de>
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

import pygtk
pygtk.require("2.0")
import gtk.gdk

import xrandr

def get_default_screen_config():
    dpy = gtk.gdk.display_get_default()
    screen = dpy.get_default_screen()
    return get_screen_config(screen)

def get_screen_config(screen):
    """Returns the XRandR screen config instance for the given gtk.gdk.Screen"""
    dpy = screen.get_display()
    dpy_url = dpy.get_name()
    count = screen.get_number()
    return xrandr.get_screen_of_display(dpy_url, count)

# vim:ts=4:sw=4:et


xrandr/cli.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# Python-XRandR provides a high level API for the XRandR extension of the
# X.org server. XRandR allows to configure resolution, refresh rate, rotation 
# of the screen and multiple outputs of graphics cards.
#
# This module provides a command line interface tool to access the
# features of XRandR.
#
# In many aspects it follows the design of the xrand tool written by
# Keith Packard.
#
# Copyright 2007 © Sebastian Heinlein <sebastian.heinlein@web.de>
# Copyright 2007 © Michael Vogt <mvo@ubuntu.com>
# Copyright 2007 © Canonical Ltd.
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

from gettext import gettext as _
import gettext
gettext.textdomain("python-xrandr")
import sys

from optparse import OptionParser

import xrandr

__version__ = "0.0.x(development)"

def main():
    parser = OptionParser(version=__version__)
    parser.add_option("-v", "",
                      action="store_true", dest="version",
                      default=False,
                      #TRANSLATORS: command line option
                      help=_("show current xrandr version"))
    parser.add_option("--verbose", "",
                      action="store_true", dest="verbose",
                      default=False,
                      #TRANSLATORS: command line option
                      help=_("show additional information"))
    parser.add_option("-s", "--size",
                      default=None,
                      action="store", type="int", dest="size",
                      #TRANSLATORS: command line option
                      help=_("set the screen reolsution to the given size id"))
    parser.add_option("-r", "--rate",
                      default=None,
                      action="store", type="int", dest="rate",
                      #TRANSLATORS: command line option
                      help=_("set the given refresh rate"))
    parser.add_option("-o", "--rotate",
                      default=None,
                      action="store", type="string", dest="rotate",
                      #TRANSLATORS: command line option
                      help=_("rotate the screen. supported values are "
                             "normal, left, inverted, right or 0, 90, 180, "
                             "270"))
    parser.add_option("--output", "",
                      default=None,
                      action="store", type="string", dest="output",
                      #TRANSLATORS: command line option
                      help=_("select an available output"))
    parser.add_option("--preferred", "",
                      action="store_true", dest="output_preferred",
                      #TRANSLATORS: command line option
                      help=_("choose the preferred resolution and rate"))
    parser.add_option("--mode", "",
                      action="store", type="int", dest="output_mode",
                      #TRANSLATORS: command line option
                      help=_("set to the mode of the given index number"))
    parser.add_option("--off", "",
                      action="store_true", dest="output_disable",
                      #TRANSLATORS: command line option
                      help=_("turn off the selected output"))
    parser.add_option("--left-of", "",
                      default=None,
                      action="store", type="string", dest="output_left",
                      #TRANSLATORS: command line option
                      help=_("move the output left of the given one"))
    parser.add_option("--right-of", "",
                      default=None,
                      action="store", type="string", dest="output_right",
                      #TRANSLATORS: command line option
                      help=_("move the output right of the given one"))
    parser.add_option("--above", "",
                      default=None,
                      action="store", type="string", dest="output_above",
                      #TRANSLATORS: command line option
                      help=_("move the output above of the given one"))
    parser.add_option("--below", "",
                      default=None,
                      action="store", type="string", dest="output_below",
                      #TRANSLATORS: command line option
                      help=_("move the output below of the given one"))
    parser.add_option("--same-as", "",
                      default=None,
                      action="store", type="string", dest="output_same",
                      #TRANSLATORS: command line option
                      help=_("move the output to the position of the "
                             "given one"))
    (options, args) = parser.parse_args()

    if xrandr.has_extension():
        if options.version:
            print "%x.%s" % xrandr.XRANDR_VERSION
            sys.exit()
        else:
            print _("XRandR %s.%s") % xrandr.XRANDR_VERSION
    else:
        print _("The XRandR extension is not available")
        sys.exit(1)

    changed_1_0 = False

    screen = xrandr.get_current_screen()

    if options.size != None:
        screen.set_size_index(options.size)
        changed_1_0 = True
    if options.rate:
        screen.set_refresh_rate(options.rate)
        changed_1_0 = True
    if options.rotate:
        if options.rotate in [_("normal"), "0"]:
            rotation = xrandr.RR_ROTATE_0
        elif options.rotate in [_("right"), "90"]:
            rotation = xrandr.RR_ROTATE_90
        elif options.rotate in [_("inverted"), "180"]:
            rotation = xrandr.RR_ROTATE_180
        elif options.rotate in [_("left"), "270"]:
            rotation = xrandr.RR_ROTATE_270
        else:
            print _("Invalid orientation")
            sys.exit(1)
        screen.set_rotation(rotation)
        changed_1_0 = True
    if changed_1_0:
        screen.apply_config()
    elif options.output:
        output = screen.get_output_by_name(options.output)
        if not output:
            print _("Output does not exist")
            sys.exit(1)
        if options.output_disable:
            output.disable()
        elif options.output_preferred:
            output.set_to_preferred_mode()
        elif options.output_mode != None:
            output.set_to_mode(options.output_mode)
        if options.output_left:
            output.set_relation(options.output_left, xrandr.RELATION_LEFT_OF)
        elif options.output_right:
            output.set_relation(options.output_right, xrandr.RELATION_RIGHT_OF)
        elif options.output_above:
            output.set_relation(options.output_above, xrandr.RELATION_ABOVE)
        elif options.output_below:
            output.set_relation(options.output_below, xrandr.RELATION_BELOW)
        elif options.output_same:
            output.set_relation(options.output_same, xrandr.RELATION_SAME_AS)
        screen.apply_output_config()
    else:
        screen.print_info(options.verbose)

if __name__ == "__main__":
    main()

# vim:ts=4:sw=4:et
